compiler/rustc_resolve/src/late.rs

```diff
@@ -3739,7 +3741,62 @@ impl<'a: 'ast, 'b, 'ast, 'tcx> LateResolutionVisitor<'a, 'b, 'ast, 'tcx> {
             }
 
-            // Make sure `A::B` in `<T as A::B>::C` is a trait item.
+            // Make sure that `A` in `<T as A>::B::C` is a trait.
+            //
+            // This is best-effort, for diagnostics, as the subsequent
+            // call to `smart_resolve_path_fragment` will fail even if
+            // we don't emit an error here.
+            let trait_path = &path[..qself.position];
+            match self.resolve_path(trait_path, Some(TypeNS), None) {
+                PathResult::NonModule(res) => {
+                    match res.full_res() {
+                        Some(Res::Def(DefKind::Trait, _)) => unreachable!(),
+                        Some(Res::Def(def_kind, def_id)) => {
+                            return Err(Spanned {
+                                node: ResolutionError::NotATrait {
+                                    name: Segment::names_to_string(trait_path),
+                                    article_and_kind: Some((
+                                        def_kind.article(),
+                                        def_kind.descr(def_id),
+                                    )),
+                                },
+                                span: trait_path[qself.position - 1].ident.span,
+                            });
+                        }
+                        Some(Res::SelfTyAlias { .. } | Res::SelfTyParam { .. }) => {
+                            return Err(Spanned {
+                                node: ResolutionError::NotATrait {
+                                    name: Segment::names_to_string(trait_path),
+                                    article_and_kind: Some(("a", "type")),
+                                },
+                                span: trait_path[qself.position - 1].ident.span,
+                            });
+                        }
+                        // FIXME(compiler-errors): Should we raise an error for `None`
+                        // resolution here, since traits aren't type-dependent?
+                        _ => {}
+                    }
+                }
+                PathResult::Module(ModuleOrUniformRoot::Module(mod_data)) => match mod_data.kind {
+                    ModuleKind::Def(DefKind::Trait, _, _) => {}
+                    ModuleKind::Def(def_kind, def_id, _) => {
+                        return Err(Spanned {
+                            node: ResolutionError::NotATrait {
+                                name: Segment::names_to_string(trait_path),
+                                article_and_kind: Some((
+                                    def_kind.article(),
+                                    def_kind.descr(def_id),
+                                )),
+                            },
+                            span: trait_path[qself.position - 1].ident.span,
+                        });
+                    }
+                    _ => {}
+                },
+                _ => {}
+            }
```

Comment:
Could you use `smart_resolve_path_fragment` with `PathSource::Trait` here instead of hand-rolling all this stuff?
Almost exactly the same thing is done a few lines below, but for `PathSource::TraitItem`s.